File: Arjuna/versions/PR3.3_97_03_27/CrashRecovery/Lib/Dozer.cc
/*
 * Copyright (C) 1993, 1994
 *
 * Department of Computing Science,
 * The University,
 * Newcastle upon Tyne,
 * UK.
 *
 * $Id: Dozer.cc,v 1.1 1994/08/02 11:02:29 ndbi Exp $
 */

#ifndef UNISTD_H_
#  include <System/unistd.h>
#endif

#ifndef TIME_H_
#  include <System/time.h>
#endif

#if defined(DEBUG) && !defined(DEBUG_H_)
#  include <Common/Debug.h>
#endif

#ifndef DOZER_H_
#  include <Dozer.h>
#endif

/*
 * Class: Dozer
 * Clever sleep with exponential backoff and possible threshold function
 * firing
 */

//-Public-Functions------------------------------------------------------------

Dozer::Dozer(ostream& strm)
            : outStrm(strm),
              baseSleepTime(1),
              currentSleepTime(1),
              threshold(-1),
              noTries(0),
              firedWarning(FALSE),
              thresholdFunction(0)
{
#ifdef DEBUG
    debug_stream << CONSTRUCTORS << FAC_CRASH_RECOVERY << VIS_PUBLIC;
    debug_stream << "Dozer::Dozer()" << endl;
#endif
}

Dozer::Dozer(ostream& strm, int initialSleepTime)
            : outStrm(strm),
              baseSleepTime(initialSleepTime),
              currentSleepTime(initialSleepTime),
              threshold(-1),
              noTries(0),
              firedWarning(FALSE),
              thresholdFunction(0)
{
#ifdef DEBUG
    debug_stream << CONSTRUCTORS << FAC_CRASH_RECOVERY << VIS_PUBLIC;
    debug_stream << "Dozer::Dozer(" << initialSleepTime << ")" << endl;
#endif
}

Dozer::~Dozer()
{
#ifdef DEBUG
    debug_stream << DESTRUCTORS << FAC_CRASH_RECOVERY << VIS_PUBLIC;
    debug_stream << "Dozer::~Dozer()" << endl;
#endif
}

void Dozer::doze()
{
    // Check threshold

    if ( (noTries == threshold) && (thresholdFunction) )
    {
        if ( (*thresholdFunction)() )
        {
             // if thresholdFunction returns TRUE then reset noTries
             noTries = 0;
        }
        firedWarning = TRUE;
    }
    else
    {
        sleep(currentSleepTime);
        noTries++;
        currentSleepTime = 2 * currentSleepTime;
    }
}

void Dozer::reset()
{
    noTries = 0;
    currentSleepTime = baseSleepTime;
    firedWarning = FALSE;
}

void Dozer::reset(int newBase)
{
    currentSleepTime = baseSleepTime = newBase;
    noTries = 0;
    currentSleepTime = baseSleepTime;
    firedWarning = FALSE;
}

void Dozer::setThreshold(int newThreshold)
{
    threshold = newThreshold;
}

void Dozer::setWarningFunction(warningFunction func)
{
    thresholdFunction = func;
}

Boolean Dozer::fired()
{
    return(firedWarning);
}

void Dozer::sleep(int sleepTime)
{
    time_t theTime = ::time(0);
    outStrm << "Dozer: Sleeping for " << sleepTime << " secs. Its now " 
            << ctime(&theTime) << flush;
    ::sleep(sleepTime);
    theTime = ::time(0);
    outStrm << "Dozer: Awake at " << ctime(&theTime) << flush;
}

File: JTSArjuna/JTSArjuna-2.0_Alpha_2_0/JavaCosServices/JavaCosRecovery/classes/com/arjuna/JavaCosRecovery/CrashMan/Dozer.java
/*
 * Copyright (C) 1996, 1997, 1998,
 *
 * Department of Computing Science,
 * University of Newcastle upon Tyne,
 * Newcastle upon Tyne,
 * UK.
 *
 * $Id: Dozer.java,v 1.2 1998/07/06 10:49:51 nmcl Exp $
 */

package com.arjuna.JavaCosRecovery.CrashMan;

import java.io.PrintStream;
import java.util.Date;
import java.lang.InterruptedException;

/*
 * Default visibility.
 */

/*
 * Class: Dozer
 * Clever sleep with exponential backoff and possible threshold function
 * firing
 */

class Dozer
{
    
public Dozer (PrintStream strm)
    {
	outStrm = strm;
	baseSleepTime = 1;
	currentSleepTime = 1;
	threshold = -1;
	noTries = 0;
	firedWarning = false;
	thresholdObject = null;
    }

public Dozer (PrintStream strm, int initialSleepTime)
    {
	outStrm = strm;	
	baseSleepTime = initialSleepTime;
	currentSleepTime = initialSleepTime;
	threshold = -1;
	noTries = 0;
	firedWarning = false;
	thresholdObject = null;
    }
    
public void doze ()
    {
	// Check threshold

	if ( (noTries == threshold) && (thresholdObject != null) )
	{
	    if (thresholdObject.reached())
	    {
		// if thresholdObject returns true then reset noTries
		noTries = 0;
	    }
	    
	    firedWarning = true;
	}
	else
	{
	    sleep(currentSleepTime);
	    noTries++;
	    currentSleepTime = 2 * currentSleepTime;
	}
    }
        
public void reset ()
    {
	noTries = 0;
	currentSleepTime = baseSleepTime;
	firedWarning = false;
    }

public void reset (int newBase)
    {
	currentSleepTime = baseSleepTime = newBase;
	noTries = 0;
	currentSleepTime = baseSleepTime;
	firedWarning = false;
    }

public void setThreshold (int newThreshold)
    {
	threshold = newThreshold;
    }

public void setThresholdObject (ThresholdObject obj)
    {
	thresholdObject = obj;
    }

public boolean fired ()
    {
	return firedWarning;
    }
    
private void sleep (int sleepTime)
    {
	Date date = new Date();

	outStrm.println("Dozer: Sleeping for "+sleepTime+" secs. Its now " 
			+date.toGMTString());

	date = null;

	try
	{
	    Thread.sleep(sleepTime*1000);
	}
	catch (InterruptedException e)
	{
	}

	date = new Date();
	
	outStrm.println("Dozer: Awake at "+date.toGMTString());
    }
    
private PrintStream outStrm;
private int baseSleepTime;
private int currentSleepTime;
private int threshold;
private int noTries;
private boolean firedWarning;
private ThresholdObject thresholdObject;

}
